aidial_adapter_openai/gpt4_vision/chat_completion.py

```diff
@@ -0,0 +1,374 @@
+import mimetypes
+from typing import (
+    Any,
+    AsyncIterator,
+    Dict,
+    List,
+    Optional,
+    Tuple,
+    TypeVar,
+    cast,
+)
+
+import aiohttp
+from fastapi.responses import JSONResponse, Response, StreamingResponse
+
+from aidial_adapter_openai.gpt4_vision.gpt4_conversion import (
+    convert_gpt4v_to_gpt4_chunk,
+)
+from aidial_adapter_openai.gpt4_vision.image_tokenizer import tokenize_image
+from aidial_adapter_openai.gpt4_vision.messages import (
+    create_image_message,
+    create_text_message,
+)
+from aidial_adapter_openai.utils.exceptions import HTTPException
+from aidial_adapter_openai.utils.image_data_url import ImageDataURL
+from aidial_adapter_openai.utils.log_config import logger
+from aidial_adapter_openai.utils.sse_stream import (
+    parse_openai_sse_stream,
+    to_openai_sse_stream,
+)
+from aidial_adapter_openai.utils.storage import (
+    FileStorage,
+    download_file_as_base64,
+)
+from aidial_adapter_openai.utils.streaming import (
+    create_error_response,
+    generate_stream,
+    map_stream,
+    prepend_to_stream,
+)
+from aidial_adapter_openai.utils.text import format_ordinal
+from aidial_adapter_openai.utils.tokens import Tokenizer
+
+# The built-in default max_tokens is 16 tokens,
+# which is too small for most image-to-text use cases.
+DEFAULT_MAX_TOKENS = 128
```

Comment:
It doesn't seem right to change the default but maybe it will be more practical.

Reply:
16 tokens is too little for any practical uses in chat